apollo-federation/src/sources/connect/models/http_json_transport.rs

```diff
@@ -0,0 +1,321 @@
+use http::Uri;
+
+use crate::error::FederationError;
+use crate::sources::connect::JSONSelection;
+use crate::sources::connect::Namespace;
+use crate::sources::connect::PathSelection;
+use crate::sources::connect::StringTemplate;
+use crate::sources::connect::header::HeaderValue;
+use crate::sources::connect::json_selection::ExternalVarPaths;
+use crate::sources::connect::spec::ConnectHTTPArguments;
+use crate::sources::connect::spec::SourceHTTPArguments;
+use crate::sources::connect::spec::schema::HEADERS_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::HTTP_HEADER_MAPPING_FROM_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::HTTP_HEADER_MAPPING_NAME_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::HTTP_HEADER_MAPPING_VALUE_ARGUMENT_NAME;
+use crate::sources::connect::spec::versions::AllowedHeaders;
+use crate::sources::connect::string_template;
+use crate::sources::connect::variable::VariableReference;
+
+#[derive(Clone, Debug)]
+pub struct HttpJsonTransport {
+    pub source_url: Option<Uri>,
+    pub connect_template: StringTemplate,
+    pub method: HTTPMethod,
+    pub headers: IndexMap<HeaderName, HeaderSource>,
+    pub body: Option<JSONSelection>,
+}
+
+impl HttpJsonTransport {
+    pub(crate) fn from_directive(
+        http: &ConnectHTTPArguments,
+        source: Option<&SourceHTTPArguments>,
+    ) -> Result<Self, FederationError> {
+        let (method, connect_url) = if let Some(url) = &http.get {
+            (HTTPMethod::Get, url)
+        } else if let Some(url) = &http.post {
+            (HTTPMethod::Post, url)
+        } else if let Some(url) = &http.patch {
+            (HTTPMethod::Patch, url)
+        } else if let Some(url) = &http.put {
+            (HTTPMethod::Put, url)
+        } else if let Some(url) = &http.delete {
+            (HTTPMethod::Delete, url)
+        } else {
+            return Err(FederationError::internal("missing http method"));
+        };
+
+        #[allow(clippy::mutable_key_type)]
+        // HeaderName is internally mutable, but we don't mutate it
+        let mut headers = http.headers.clone();
+        for (header_name, header_source) in
+            source.map(|source| &source.headers).into_iter().flatten()
+        {
+            if !headers.contains_key(header_name) {
+                headers.insert(header_name.clone(), header_source.clone());
+            }
+        }
+
+        Ok(Self {
+            source_url: source.map(|s| s.base_url.clone()),
+            connect_template: connect_url.parse().map_err(|e: string_template::Error| {
+                FederationError::internal(format!(
+                    "could not parse URL template: {message}",
+                    message = e.message
+                ))
+            })?,
+            method,
+            headers,
+            body: http.body.clone(),
+        })
+    }
+
+    pub(crate) fn label(&self) -> String {
+        format!("http: {} {}", self.method.as_str(), self.connect_template)
```

Comment:
`self.method` inside the format doesn't need the `as_str()` as it already implements that in `Display`:

```rust
impl Display for HTTPMethod {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}
```